### PR TITLE
[quant] Input-Weight Equalization - edge cases

### PR DESCRIPTION
**Pull request by angelayi**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #62367

Summary: Addresses the edge cases where there exists skippable nodes like relu and pooling, and addresses the case where there are nodes that are being quantized but not equalized.

Modifications:
- When we are finding the next equalization scale for additional scaling in the case of connected layers
    - An example is: `x -> conv1 -> relu -> maxpool -> conv2`
    - After prepare: `x -> inp_quant_obs1 -> inp_eq_obs1 -> conv1 -> relu -> out_quant_obs1 -> maxpool -> inp_eq_obs2 -> conv2 -> out_quant_obs2`
    - At the starting node (ex. `conv1`), we will traverse through the graph skipping all relu/batchnorm nodes until we reach a call_module node. If it is a quantization observer (`out_quant_obs2`), we will continue.
    - We will then continue traversing the graph (skipping the same nodes) until we reach another call_module node. If it is an equalization observer (`inp_eq_obs2`), this means we have two connected layers that need to be equalized.
    - So, we will scale the current op node (`conv1`) with the next equalization scale.

- Finding the previous op node to see if we need to remove the current equalization observer or replace it with a mul node
    - Using the same example as above: `x -> conv1 -> relu -> maxpool -> conv2`
    - After prepare: `x -> inp_quant_obs1 -> inp_eq_obs1 -> conv1 -> relu -> out_quant_obs1 -> maxpool -> inp_eq_obs2 -> conv2 -> out_quant_obs2`
    - At the starting node (ex. `conv2`), we will traverse backwards through the graph skipping all relu/batchnorm nodes until we find a call_module node. If it is a quantization observer (`out_quant_obs1`), we will continue.
    - We will then continue traversing the graph (skipping the same nodes) until we reach another call_module node.
    - If this node (`conv1`) is being equalized, this means we are in the case where there are two connected layers that are being equalized, so we will remove the current equalization observer.
    - If this node is not being equalized, then we will replace the equalization observer with a mul node like how we did before.

Test Plan: `python test/test_quantization.py TestEqualizeFx`

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29989948](https://our.internmc.facebook.com/intern/diff/D29989948)

### CODE REVIEW

**Comment by angelayi** [comment]
@angelayi has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29989948).

File: torch/quantization/fx/_equalize.py [hunk 1]
```diff
@@ -318,22 +414,12 @@ def maybe_get_next_input_eq_obs(node: Node, modules: Dict[str, nn.Module]) -> Op
...
-        maybe_relu_node = maybe_get_next_module(node, modules, target_functional_type=F.relu)
-
     # Locate the following output observer if it exists.
-    # We will skip the relu node if it exists.
-    maybe_obs_node = (
-        maybe_get_next_module(node, modules, ObserverBase)
-        if maybe_relu_node is None
-        else maybe_get_next_module(maybe_relu_node, modules, ObserverBase)
-    )
+    maybe_obs_node = maybe_get_following_observer(node, modules, ObserverBase)
```
Comment (supriyar) [inline torch/quantization/fx/_equalize.py:418]: Why do we need to find `maybe_obs_node` here? Can we directly find `maybe_get_following_observer(node, modules, _InputEqualizationObserver)`?

File: torch/quantization/fx/_equalize.py [hunk 2]
```diff
@@ -410,7 +496,12 @@ def scale_weight_node(
 
     # Multiply the weights row wise by the next equalization scale
     # Reshape the equalization scale so that we can multiply it to the weight along axis=0
-    next_equalization_scale_reshaped = reshape_scale(next_equalization_scale, 0, weight)
+    try:
+        next_equalization_scale_reshaped = reshape_scale(next_equalization_scale, 0, scaled_weight)
+    except RuntimeError:
+        raise ValueError(
+            "Input-weight equalization is unable to support general reshapes within models."
+        )
```
Comment (supriyar) [inline torch/quantization/fx/_equalize.py:504]: Would be useful to add a more descriptive message here saying what the mismatch was

File: torch/quantization/fx/quantization_patterns.py [hunk 3]
```diff
@@ -82,11 +83,20 @@ def _maybe_get_last_node_only_observer(
         If the last node of the pattern is observed, return the observer
         instance. Otherwise, return None.
         """
-        for maybe_obs_node, _ in self.last_node.users.items():
+        last_node = self.last_node
+
+        # Handle the additional mul node that gets added during equalization if
```
Comment (supriyar) [inline torch/quantization/fx/quantization_patterns.py:88]: Why is this a special case here?

Comment (angelayi) [reply torch/quantization/fx/quantization_patterns.py:88]: The reason why there is a special casing when there is a quantized but not equalized node followed by an equalized node. 

For example, let's say linear1 is not being equalized but linear2 is being equalized. The following graph is what we get after prepare:
```
x -> inp_quant_obs1 -> linear1 -> out_quant_obs1 -> inp_eq_obs2 -> linear2 -> out_quant_obs2
```

After the equalization functions, we have:
```
x -> inp_quant_obs1 -> linear1 -> mul -> out_quant_obs1 (scaled) -> linear2 -> out_quant_obs2
```

In this current function (`maybe_get_last_node_observer`), it originally only looped through the op node's users to find its output quantization observer. However after applying the equalization nodes, linear1's output quantization observer no longer appears directly after linear1, so I added some extra casing.

It should not interfere with the case where the model already has an op node, because in that case I think there would be a output quantization observer inserted before the mul node:
```
x -> inp_quant_obs1 -> linear1 -> out_quant_obs1 -> mul -> out_quant_obs2
```

File: torch/quantization/fx/_equalize.py (hunk 1 above)
Comment (angelayi) [reply torch/quantization/fx/_equalize.py:418]: The way the check current works is that it will look for the following call_module node and see if it's the specific given type. However, there should always be a quantization observer in the middle between the op_node and equalization observer, which would result in `maybe_get_following_observer(node, modules, _InputEqualizationObserver)` returning None.

So I first found the quantization observer, and then looked for the equalization observer from the quantization observer.

File: torch/quantization/fx/quantization_patterns.py (hunk 3 above)
Comment (supriyar) [reply torch/quantization/fx/quantization_patterns.py:88]: What about the case where linear1 isn't being quantized and we have a mul node after it. Would we end up calling this function in that case?

@jerryzh168 could you help take a look at this change and see if it is okay for the non-equalization use case?

File: torch/quantization/fx/_equalize.py [hunk 4]
```diff
@@ -298,6 +322,78 @@ def maybe_get_weight_eq_obs_node(op_node: Node, modules: Dict[str, nn.Module]) -
...
+    returning the quantization observer.
+    If we are looking for the equalization observer, we want to start from the
+    given node (inp_quant_obs) and traverse forwards skipping the pooling layer,
+    and returning the equalization observer.
+    """
+
+    temp_node = node
+    while len(temp_node.users.items()) > 0:
+        possible_node = None
+        for user, _ in temp_node.users.items():
```
Comment (jerryzh168) [inline torch/quantization/fx/_equalize.py:346]: nit: does `for user in temp_node.users` work?

File: torch/quantization/fx/_equalize.py [hunk 5]
```diff
@@ -298,6 +322,78 @@ def maybe_get_weight_eq_obs_node(op_node: Node, modules: Dict[str, nn.Module]) -
             return node_arg
     return None
 
+def maybe_get_following_observer(
+    node: Node,
+    modules: Dict[str, nn.Module],
+    target_observer_type: Any = object,
+) -> Optional[Node]:
+    """ Loops through the graph forwards and finds the next observer node with
```
Comment (jerryzh168) [inline torch/quantization/fx/_equalize.py:330]: if there are multiple paths, looks like this will just follow through one path and find the first observer, might be good to document this I think.

File: torch/quantization/fx/_equalize.py [hunk 6]
```diff
@@ -298,6 +322,78 @@ def maybe_get_weight_eq_obs_node(op_node: Node, modules: Dict[str, nn.Module]) -
...
+        - The previous quantization observer node that appears before the
+        previous op node if it exists
+    """
+
+    inp_quant_obs_node = None
+    prev_node = None
+
+    temp_node = node
+    while len(temp_node.args) > 0:
+        possible_node = temp_node.args[0]
```
Comment (jerryzh168) [inline torch/quantization/fx/_equalize.py:381]: I feel this loop is a little bit confusing, can we just assign `temp_node = temp_node.args[0]`here and redo the condition checks below?

looks like in all three branches we can do `temp_node = temp_node.args[0]`

File: torch/quantization/fx/quantization_patterns.py (hunk 3 above)
Comment (jerryzh168) [reply torch/quantization/fx/quantization_patterns.py:88]: right, there is a chance that the mul following linear is not quantized, for example user explicitly configures that to be not quantized through qconfig_dict, I feel this is probably a bit too hacky, let's think about if we have a different workaround

Comment (angelayi) [reply torch/quantization/fx/quantization_patterns.py:88]: > What about the case where linear1 isn't being quantized and we have a mul node after it. Would we end up calling this function in that case?

If linear1 isn't being quantized, then I think it wouldn't enter this function.

> the mul following linear is not quantized, for example user explicitly configures that to be not quantized through qconfig_dict.

In this case, I think the structure of the prepared graph would still be, so the mul would not appear between the linear and output quantization observer if I'm not mistaken:
```
x -> inp_quant_obs1 -> inp_eq_obs -> linear1 -> out_quant_obs1 -> mul
```

Comment (jerryzh168) [reply torch/quantization/fx/quantization_patterns.py:88]: what if linear is also not quantized and the op following mul is quantized? would we have 
```
linear1 -> mul -> observer -> conv -> ...
```   and when we call `_maybe_get_last_node_only_observer` it would return the observer?

Comment (angelayi) [reply torch/quantization/fx/quantization_patterns.py:88]: I see, I understand that case now. Could we check to see if the mul operator contains a node with the name "equalization_scale" as an argument? Or I could modify the name of the mul node for equalization to be like "mul_for_equalization" and check if the mul operator found contains "equalization" in the name?

**Comment by angelayi** [comment]
@angelayi has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29989948).

File: torch/quantization/fx/quantization_patterns.py (hunk 3 above)
Comment (supriyar) [reply torch/quantization/fx/quantization_patterns.py:88]: > Could we check to see if the mul operator contains a node with the name "equalization_scale" as an argument?

This seems okay to me. Can we add an argument to this function to denote equalization is enabled, in which case it checks for this argument for a `mul`?

**Comment by angelayi** [comment]
@angelayi has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29989948).

File: torch/quantization/fx/quantization_patterns.py [hunk 7]
```diff
@@ -1607,7 +1638,8 @@ def convert(self,
                 node_name_to_scope: Dict[str, Tuple[str, type]],
                 load_arg: Callable,
                 is_reference: bool = False,
-                convert_custom_config_dict: Dict[str, Any] = None) -> Node:
+                convert_custom_config_dict: Dict[str, Any] = None,
+                equalization_enabled: bool = False) -> Node:
```
Comment (supriyar) [inline torch/quantization/fx/quantization_patterns.py:1642]: Do we need to add this argument to add the instances of `convert`, given that it has a default value?

**Review by supriyar — APPROVED** [review]
Looks good, please take a look at the comment before landing

**Comment by angelayi** [comment]
@angelayi has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29989948).

**Comment by angelayi** [comment]
@angelayi has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29989948).

File: torch/quantization/fx/quantization_patterns.py [hunk 8]
```diff
@@ -76,17 +77,33 @@ def __init__(self, node: Node, modules: Dict[str, torch.nn.Module]):
...
+
+        if equalization_enabled:
+            # Handle the additional mul node that gets added during equalization if
+            # it exists.
+            maybe_eq_node = maybe_get_next_module(last_node, modules, target_functional_type=torch.mul)
+            if maybe_eq_node is not None:
+                # Check that the mul node contains an 'equalization_scale' as an
+                # argument to ensure that this mul node is used for equalization.
+                for arg in maybe_eq_node.args:
+                    if isinstance(arg, Node) and "equalization_scale" in arg.name:
```
Comment (jerryzh168) [inline torch/quantization/fx/quantization_patterns.py:97]: actually Node name is a fragile thing to check I think, is it possible to check for something else?

Comment (angelayi) [reply torch/quantization/fx/quantization_patterns.py:97]: Could we check with node.target?

**Comment by angelayi** [comment]
@angelayi has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29989948).

**Comment by supriyar** [comment]
@angelayi is the mypy failure related to the changes?

**Comment by angelayi** [comment]
@angelayi has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29989948).
